homeworks/01-bubble-sort/src/App.tsx

```diff
@@ -0,0 +1,189 @@
+import { ChangeEvent, Component } from "react";
+
+import Controllers from "./Controllers";
+import Settings from "./Settings";
+import Bar from "./Bar";
+
+const generateArray = (length: number = 30): number[] =>
+  Array.from(Array(length))
+    .map((_, index) => (index + 1) * 3)
+    .sort(() => Math.random() - 0.5);
+
+const swap = (arr: number[], a: number, b: number): void => {
+  const temp = arr[a];
+  arr[a] = arr[b];
+  arr[b] = temp;
+};
+
+const delay = (timer: number): Promise<void> =>
+  new Promise((resolve) => setTimeout(resolve, 400 - timer));
+const sortArray = async (
+  array: number[],
+  updateState: <K extends keyof State>(state: Pick<State, K>) => void,
+  settings: { speed: number; pause: boolean; isActive: boolean }
+) => {
+  const copyArr = [...array];
+  const arrLength = copyArr.length;
+
+  function pause(): Promise<void> {
+    return new Promise((resolve) => {
+      if (settings.pause) {
+        return setTimeout(() => resolve(pause()), 300);
+      } else {
+        return resolve();
+      }
+    });
+  }
+
+  for (let i = 0; i < arrLength; i++) {
+    for (let j = 0; j < arrLength - i - 1; j++) {
+      if (settings.pause) {
+        await pause();
+      }
+      if (!settings.isActive) {
+        break;
+      }
+
+      const a = copyArr[j];
+      const b = copyArr[j + 1];
+      if (a > b && settings.isActive) {
+        swap(copyArr, j, j + 1);
+        updateState({ renderData: [...copyArr] });
+        await delay(settings.speed);
+      }
+    }
+  }
+
+  if (settings.isActive) {
+    updateState({
+      sorting: false,
+      solved: true,
+    });
+  }
+};
+
+const DEFAULT_ARRAY_LENGTH = 30;
+const DEFAULT_SORT_SPEED = 50;
+
+type State = {
+  arraylength: number;
+  speed: number;
+  data: number[];
+  renderData: number[];
+  pause: boolean;
+  sorting: boolean;
+  solved: boolean;
+};
+
+class App extends Component<{}, State> {
+  settings = {
+    speed: DEFAULT_SORT_SPEED,
+    pause: false,
+    isActive: false,
+  };
+
+  state: State = {
+    arraylength: DEFAULT_ARRAY_LENGTH,
+    speed: DEFAULT_SORT_SPEED,
+    data: generateArray(DEFAULT_ARRAY_LENGTH),
+    renderData: [],
+    sorting: false,
+    pause: false,
+    solved: false,
+  };
+
+  componentDidMount() {
+    const { data } = this.state;
+    this.setState({ renderData: [...data] });
+  }
+
+  handleNewSet = () => {
+    this.settings.isActive = false;
+    this.settings.pause = false;
+    const newArray = generateArray();
```

Comment:
When I click `New set`, it's not taken into account the `Number of element` value.

Also, the scale becomes incorrect.

![image](https://user-images.githubusercontent.com/24900675/115211073-e2645000-a107-11eb-807b-ddf50742dfb2.png)

![image](https://user-images.githubusercontent.com/24900675/115211114-ee501200-a107-11eb-909f-ddc85556c326.png)